std.algorithm: {min,max}Element for a single range

wilzbach opened the pull request:
resubmisson of #4019 - as `extremum` turned out not to be very useful, I changed it to accept directly the mapping function. This is now more efficient as the mapping function is called only once :)

@andralex has already approved the name:

> Overall I think this is rather foreign compared to prevalent Phobos style. I think extremum,minElement, and maxElement with the appropriate predicates are the way to go.

Pinging @JackStouffer @burner @ntrel @klickverbot @JackStouffer

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
+
+    static if (isRandomAccessRange!Range && hasLength!Range)
+    {
+        size_t lenArray = r.length;
```
`immutable`

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
+    // init extremum
+    Unqual!(ElementType!Range) extremeElement = r.front;
+    alias mapType = Unqual!(typeof(mapFun((ElementType!Range).init)));
+    mapType extremeElementMapped = mapFun(r.front);
```
Why `mapType` and not `auto`? Same throughout.

wilzbach replied:
It could be that the type is `immutable` or `const` - hence we need to remove that -> `Unqual`

yebblies replied:
Should at least be `MapType` since it's a type.

wilzbach replied:
> Should at least be MapType since it's a type.

Ok - changed ;-)

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3035,111 @@ unittest
+A custom mapping for comparison key can be used with $(D map).
+
+Params:
+    map = custom mapping for the comparison key
```
Saying "custom accessor" would be more clear IMO

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
+
+    // allows a custom selector for comparison
+    assert([[0, 4], [1, 2]].extremum!("a[0]", "a > b") == [1, 2]);
+    assert([[0, 4], [1, 2]].extremum!("a[1]", "a > b") == [0, 4]);
```
Why can't the two template parameters be combined into `"a[1] > b[1]"`?

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3035,111 @@ unittest
+    import std.range: enumerate, iota;
+    assert([2, 1, 4, 3].maxElement == 4);
+    // allows to get the index too
+    assert([2, 1, 4, 3].enumerate.maxElement!"a.value" == tuple(2, 4));
```
Needs another example of a custom accessor, something like the assert on line 1279

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3035,111 @@ unittest
+
+    assert([2, 1, 4, 3].minElement == 1);
+    // allows to get the index too
+    assert(iota(1, 10).enumerate.minElement!"a.value" == tuple(0, 1));
```
Same thing here, needs another example

JackStouffer commented:
This also needs tests using the dummy ranges, as you have no tests currently with input ranges.

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -59,6 +59,12 @@ $(T2 minCount,
         $(D maxCount([2, 4, 1, 4, 1])) returns $(D tuple(4, 2)).)
+$(T2 minElement,
+        Selects the minimal element of a range.
+        $(D minElement([3, 4, 1, 2])) returns $(D 1).)
```
Should use the new \`\` style where ever possible. Same throughout.

JackStouffer commented:
Also, I forgot to mention that this should also have unit tests with `@nogc @safe nothrow pure`.

wilzbach commented:
> This also needs tests using the dummy ranges, as you have no tests currently with input ranges.
> Also, I forgot to mention that this should also have unit tests with @nogc @safe nothrow pure.

Thanks a lot for your review @JackStouffer!
All comments besides `mapType` were addressed :)

ZombineDev commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
+    alias selectorFun = binaryFun!selector;
+
+    // init extremum
+    Unqual!(ElementType!Range) extremeElement = r.front;
```
Nitpick - how about:
```
alias Element = ElementType!Range;
alias MapType = Unqual!(typeof(mapFun(Element.init)));

Unqual!Element extremeElement = r.front;
MapType extremeElementMapped = mapFun(extremeElement);
```
(Evaluate `r[0]` once and make the code a bit more DRY).

ZombineDev commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
+    static if (isRandomAccessRange!Range && hasLength!Range)
+    {
+        immutable size_t lenArray = r.length;
+        for (size_t i = 1; i < lenArray; i++)
```
`foreach (const i; 1 .. r.length)`

ZombineDev replied:
No, actually I think that merging the input and random-access range paths has no disadvantages:
```
alias Element = ElementType!Range;
alias MapType = Unqual!(typeof(mapFun(Element.init)));

// init extremum
Unqual!Element extremeElement = r.front;
MapType extremeElementMapped = mapFun(extremeElement);

r.popFront();
foreach (const element; r)
{
    MapType elementMapped = mapFun(element);
    if (selectorFun(elementMapped, extremeElementMapped))
    {
        extremeElement = element;
        extremeElementMapped = elementMapped;
    }
}

return extremeElement;
```

ZombineDev replied:
The above will also evaluate `r[i]` / `r.front` only once.

wilzbach replied:
@ZombineDev I totally agree with you, but I had a look at other functions in this module and a lot of them (e.g. commonPrefix, countUntil, find, findSplit, {min,max}Count, {min,max}Pos, startsWith) had those two cases, so it seems to matter in terms of speed.

ntrel commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
     }
 }
 
+/**
```
Need to undocument this ;-)

wilzbach replied:
Why? It's set to `private`, so it will never be exposed to the user and for Phobos devs looking at the source it's nice to have some documentation.

JackStouffer replied:
I agree, all functions should have ddoc.

ntrel replied:
OK, could leave as-is. (I wasn't suggesting removing the comment, just the doccommentness).

ntrel commented on std/algorithm/searching.d:
```diff
@@ -1204,6 +1210,75 @@ bool endsWith(alias pred, R)(R doesThisEnd)
+        immutable size_t lenArray = r.length;
+        for (size_t i = 1; i < lenArray; i++)
+        {
+            MapType curElement = mapFun(r[i]);
```
I think `curElement` should be named e.g. `mapElement.` It's not actually an element in `r`. Given clearer naming, personally i would remove the `MapType` alias, AFAICT `auto` could be used instead.

ntrel replied:
OK, maybe we do need the alias for the `Unqual.`

ntrel commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3035,152 @@ unittest
 
+/**
+Iterates the passed range and returns the minimal element.
+A custom mapping for comparison key can be used with `map`.
```
Suggest: "A custom mapping function can be passed to `map`."

ntrel commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3035,152 @@ unittest
+Returns: The minimal element of the passed-in range.
+
+See_Also:
+    $(XREF min)
```
Just wondering if this is going to link correctly as `min` is in another module?

wilzbach replied:
Thanks - nope you were right, it doesn't resolve correctly. Fixed.

ntrel commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3035,152 @@ unittest
+///
+@safe pure unittest
+{
+    import std.range: enumerate, iota;
```
Don't need iota here.

ntrel commented:
LGTM besides minor comments.

wilzbach commented:
> LGTM besides minor comments.

Thanks a lot for reviewing this @ntrel - I just addressed your comments.
The only open question so far is whether the specialization for a RandomAccessRange is useful in terms of speed.

nordlow commented:
I don't think having an assert doing the emptyness-checking is robust enough. An `enforce` or explicit throw should be used instead. Or should we rely on the `.front`-call in these cases?

Further, I would like an optional second parameter, `seed`, (like is used in `std.algorithm.iteration.reduce`) that elides this throw-logic for flexibility, "potentially" making things `nothrow`.

DmitryOlshansky commented:
> Is having an assert do the emptyness-checking robust enough? Is this the Phobos-way?

AFAIK asserting with ranges is the style.

9il commented:
> AFAIK asserting with ranges is the style.

We need to fix `std.numeric` thought

nordlow commented:
@9il What do you mean by: fixing `std.numeric`?

9il commented:
> @9il What do you mean by: fixing std.numeric?

It uses `enforce` instead of `assert`. It should be changes to `asserts`

wilzbach commented:
> Further, I would like an optional second parameter, seed, 

@nordlow I added support for a seed element - could you please take a look? :)

Any other objections?

nordlow commented:
@wilzbach , looks good.

Perhaps this implementation should , as `reduce` does in `std.algorithm.iteration`, allow seed type to be different from `ElementType!R`. The CT-logic for handling this is quite complicated. Perhaps parts of it could be reused? Thanks.

ntrel commented:
Note that `seed` can be mimicked with `r.chain(only(seed)).maxElement`. Or with a wrapper it could be `r.orElse(seed).maxElement`.

wilzbach commented:
>  Perhaps parts of it could be reused but in a clearer way?

@nordlow done with `CommonType` + added a test for it.
@ntrel I am not convinced for neither side (pro `seed`, pro `chain`) atm - the advantage of `seed` is clear, but as long as we just use `assert`, it's still `nothrow`.
If we go the `seed` route, I guess there are a couple of more algorithms in Phobos that could use such a seed.

> It uses enforce instead of assert. It should be changes to asserts

+1

nordlow commented:
Looks good!

Now to play devils advocate: could we supply `minmaxElement` aswell:

C++ has it:

http://en.cppreference.com/w/cpp/algorithm/minmax_element

:)

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3063,186 @@ unittest
+
+    // with all dummy ranges
+    import std.internal.test.dummyrange;
+    foreach (DummyType; AllDummyRanges) {
```
brace style

JackStouffer commented on std/algorithm/searching.d:
```diff
@@ -2960,6 +3063,186 @@ unittest
+
+    // with all dummy ranges
+    import std.internal.test.dummyrange;
+    foreach (DummyType; AllDummyRanges) {
```
ditto

JackStouffer commented:
LGTM sans comments

wilzbach commented:
> LGTM sans comments

Addressed.

> Looks good! Now to play devils advocate: could we supply minmaxElement aswell? C++ has it:

Sure, but I submitted this as a separate PR (see #4248). The method is different anyways and `{min,max}Element` already have "name-approval".

9il commented:
Why not to use `map!(...).cache.reduce!min` instead?

wilzbach commented:
> Why not to use map!(...).cache.reduce!min instead?

Apart from the fact that `cache` doesn't provide random access which I think is faster, the main is that you would only get the mapping back, not the original element.

9il commented:
> Apart from the fact that cache doesn't provide random access which I think is faster, the main is that you would only get the mapping back, not the original element.

Why random access is needed for cumulative operation?
Random access is not faster. The good example is `Slice` and `byElement`. Or you have a good example where it faster?

9il commented:
> the main is that you would only get the mapping back, not the original element

Is this is perfomance problem?

wilzbach commented:
> Why random access is needed for cumulative operation? Random access is not faster

This has been discussed [here before](https://github.com/dlang/phobos/pull/4221#discussion_r60548784) and to be honest I don't know, but most other algos in this module use it too (e.g. commonPrefix, countUntil, find, findSplit, {min,max}Count, {min,max}Pos, startsWith). So I guess if we find that RandomAccess is not faster, we should remove the code for the other algos too.

@9il have you read this?
> the main [reason] is that you would only get the mapping back, not the original element.

see e.g.

```
assert([2, 1, 4, 3].enumerate.maxElement!"a.value" == tuple(2, 4));
```

9il commented:
> This has been discussed here before and to be honest I don't know, but most other algos in this module use it too (e.g. commonPrefix, countUntil, find, findSplit, {min,max}Count, {min,max}Pos, startsWith). So I guess if we find that RandomAccess is not faster, we should remove the code for the other algos too.

It is not faster and we can remove such optimizations. Furthermore they are significantly slowdown `ndslice`.

9il commented:
> the main [reason] is that you would only get the mapping back, not the original element.
> see e.g.
> 
> assert([2, 1, 4, 3].enumerate.maxElement!"a.value" == tuple(2, 4));

I don't understand. Could you please explain why access to preserved value is bad?

wilzbach commented:
> I don't understand. Could you please explain why access to preserved value is bad?

Oh no - you want to preserve the original element, if you just use `map` and `reduce` you will loose the tuple and thus the index:

```
assert([2, 1, 4, 3].enumerate.maxElement!"a.value" == tuple(2, 4));
assert([2, 1, 4, 3].map!"a.value".cache.reduce!max == 4);
```

9il commented:
Oh, OK. understand now)

9il commented:
thx)

9il commented:
Wait a minute please

9il commented:
```d
assert([2, 1, 4, 3].enumerate.reduce!(a.value < b.value ? a : b) == tuple(2, 4));
 ```

What wrong with this code?

9il commented:
If we are so lazy to write lambdas then just allow min/max to accept map function instead ;-)

wilzbach commented:
> What wrong with this code?

It evaluates the map function twice ;-)

> If we are so lazy to write lambdas then just allow min/max to accept map function instead ;-)

Yes I would love to see `min` working for ranges!

> As you mentioned a lot of languages (Java, C#, Python, more at Rosetta) allow min to be used with one argument and as still being a newcomer to D I feel that many people like me will just expect [1,2,3].min to work.

9il commented:
Another option is add combined `mapReduce` to the iteration module.

9il commented:
> It evaluates the map function twice ;-)

So we need combined `mapReduce`

9il commented:
But please do not rush to write it right now. We need to think how this can be generalized for multidimensional `Slice`

wilzbach commented:
btw just a quick update - @9il and I tested for a couple of array types with random access & `-release -O3 -boundscheck=off`.
-> the variant with RandomAccess is faster ([source](http://sprunge.us/DLbF)) (even for ndslice).

DmitryOlshansky commented:
@9il is this good to go?

wilzbach commented:
@DmitryOlshansky @9il what is your opinion on renaming this to `min` ? The other `min` expects two or more parameters and @GassaFM's comment about tuple expansion with a single array in the tuple is seldom and then still understandable.

9il commented:
OK, we can move forward. Looks like `ndslice` will have its own `functional`

9il commented:
@wilzbach `min` may cause bad behavior with templated code. I prefer current name

DmitryOlshansky commented:
Same here - minElement is established name, and single ard min is problematic in generic code

DmitryOlshansky commented:
Auto-merge toggled on

9il commented:
Hmm... how can I use this new functions with http://dlang.org/phobos/std_math.html#.cmp ?

wilzbach commented:
> how can I use this new functions with http://dlang.org/phobos/std_math.html#.cmp ?

For reference: see #4257 where I propose to make extremum public ;-)

nordlow commented:
Thanks!

andralex commented:
Why are minElement/maxElement not in the documentation at https://dlang.org/phobos/std_algorithm_searching.html?

wilzbach commented:
> Why are minElement/maxElement not in the documentation at https://dlang.org/phobos/std_algorithm_searching.html?

They are part of the 2.072 release (it's quite big due to the long accumulation period - April to September). Hopefully it can be released soon, but with all the regressions popping up it doesn't look good :/

andralex commented:
Thanks for the info.
